src/Features/Core/Portable/CodeFixes/CodeFixService.cs

```diff
@@ -73,72 +71,83 @@ internal partial class CodeFixService : ICodeFixService
+                document, range, GetShouldIncludeDiagnosticPredicate(document, priority),
+                includeSuppressedDiagnostics: false, priority, cancellationToken).ConfigureAwait(false);
 
+            var spanToDiagnostics = ConvertToMap(allDiagnostics);
+
+            using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
             var linkedToken = linkedTokenSource.Token;
 
-            // This flag is used by SuggestedActionsSource to track what solution is was
-            // last able to get "full results" for.
-            var isFullResult = await _diagnosticService.TryAppendDiagnosticsForSpanAsync(
-                document, range, diagnostics, cancellationToken: linkedToken).ConfigureAwait(false);
+            var spanToErrorDiagnostics = new SortedDictionary<TextSpan, List<DiagnosticData>>();
+            var spanToOtherDiagnostics = new SortedDictionary<TextSpan, List<DiagnosticData>>();
 
-            var errorDiagnostics = diagnostics.Where(d => d.Severity == DiagnosticSeverity.Error);
-            var otherDiagnostics = diagnostics.Where(d => d.Severity != DiagnosticSeverity.Error);
+            foreach (var (span, diagnostics) in spanToDiagnostics)
+            {
+                foreach (var diagnostic in diagnostics)
+                {
+                    var otherMap = diagnostic.Severity == DiagnosticSeverity.Error
+                        ? spanToErrorDiagnostics
+                        : spanToOtherDiagnostics;
 
-            // Kick off a task that will determine there's an Error Diagnostic with a fixer
-            var errorDiagnosticsTask = Task.Run(
-                () => GetFirstDiagnosticWithFixAsync(document, errorDiagnostics, range, linkedToken),
-                linkedToken);
+                    var list = otherMap.GetOrAdd(span, static _ => new List<DiagnosticData>());
+                    list.Add(diagnostic);
+                }
+            }
 
-            // Kick off a task that will determine if any non-Error Diagnostic has a fixer
-            var otherDiagnosticsTask = Task.Run(
-                () => GetFirstDiagnosticWithFixAsync(document, otherDiagnostics, range, linkedToken),
-                linkedToken);
+            var errorFixTask = Task.Run(() => GetFirstFixAsync(spanToErrorDiagnostics, cancellationToken), cancellationToken);
+            var otherFixTask = Task.Run(() => GetFirstFixAsync(spanToOtherDiagnostics, linkedToken), linkedToken);
 
             // If the error diagnostics task happens to complete with a non-null result before
             // the other diagnostics task, we can cancel the other task.
-            var diagnostic = await errorDiagnosticsTask.ConfigureAwait(false)
-                ?? await otherDiagnosticsTask.ConfigureAwait(false);
+            var collection = await errorFixTask.ConfigureAwait(false) ??
+                             await otherFixTask.ConfigureAwait(false);
             linkedTokenSource.Cancel();
```

Comment:
imo, this part of the design is unfortunate, though i get why we have it.  we basically want to return an error if we have one so that the lightbulb can be red not yello.  However, that means that even if we determine that we have some *other* fix first, we can't quickly update the UI.